### DISCUSSION TITLE
Environment that behaves like Python/IPython REPL

### DISCUSSION THREAD

**Discussion by tndrle** [tndrle/PyLuaTeX#3] [Ideas]
It would be nice to have an environment that behaves like a Python/IPython REPL (like the `pyconsole` environment of Pythontex).

**Comment by tndrle** [comment]
@jowe19 Hi, I added the REPL/console functionality in the branch `repl`. There is an example in `example/repl.tex`. If you are still interested in this, feel free to test it. I'd love to hear your feedback.

**Comment by jowe19** [comment]
Thanks, good news ! 😊 No problem for the moment, I'm going to test on several files during the week and would make a feedback.

**Comment by jowe19** [comment]
I have one question : is there possible to share variables between REPL code and classical python code ? (in `\py` command or `python` environment). Actually both seem to be independent ?

**Reply by tndrle** [reply]
> Actually both seem to be independent ?

The way it is implemented, they are not independent. They share the same session internally.

Can you show an example in which it is not working, please?

For me, the following works:
```latex
\documentclass{article}

\usepackage{pyluatex}
\usepackage{listings}
\usepackage{luacode}

\begin{luacode}
function pytypeset()
    tex.print("\\begin{lstlisting}[language=Python]")
    tex.print(pyluatex.get_last_output())
    tex.print("\\end{lstlisting}")
end
\end{luacode}

\newenvironment{pyrepl}
{\PyLTVerbatimEnv\begin{pythonrepl}}
{\end{pythonrepl}\directlua{pytypeset()}}

\begin{document}

\pyc{a = 3}

\begin{pyrepl}
b = a**2
\end{pyrepl}

\py{b}

\end{document}
```

**Reply by jowe19** [reply]
Sorry, I think that it was a bug in my tex file. All is working perfectly. 

Thanks again for this feature !

**Reply by jowe19** [reply]
Hi, 
Actually, when there is an error in a REPL code, the typesetting doesn't work. But sometimes I would like to be able to show classical python errors to my students. For exemple, changing a `string` or `tuple` (see below). 
<img width="464" alt="Capture d’écran 2021-11-11 à 15 13 32" src="https://user-images.githubusercontent.com/19306526/141312758-ee9f1f94-a10b-45a6-abd8-297e263a4375.png">

Is there a way to compile anyway, while showing the error message ?

**Reply by tndrle** [reply]
Hi,
there is a global package option `ignoreerrors` for this use case (see https://github.com/tndrle/PyLuaTeX#package-options).
You can also set and unset the option locally anywhere in the document with the `pyoption` command (see https://github.com/tndrle/PyLuaTeX#macros), e.g. `\pyoption{ignoreerrors}{true}` and `\pyoption{ignoreerrors}{false}`.

Does that work for you?

**Reply by jowe19** [reply]
Hi, 

Thanks, that was exactly what I needed.

**Reply by jowe19** [reply]
Just a detail : actually, when I execute a function in a structure `for` as below
<img width="818" alt="Capture d’écran 2021-11-21 à 18 16 49" src="https://user-images.githubusercontent.com/19306526/142772185-c59b1673-47c5-46b5-9d59-01e80a03f85d.png">

```
\documentclass{article}

\usepackage{pyluatex}
\usepackage{listings}
\usepackage{xcolor}
\lstset{
    language=Python,
    breaklines=true,
    framesep=1ex,
    frame=lrtb,
    framerule=0pt,
    numbers=none,
    basicstyle=\ttfamily,
    keywordstyle=\bfseries\color{green!40!black},
    stringstyle=\bfseries\color{red!80!black},
    identifierstyle=\color{blue},
    backgroundcolor=\color{gray!10!white},
}
\usepackage{luacode}

\begin{luacode}
function pytypeset()
    tex.print("\\begin{lstlisting}")
    tex.print(pyluatex.get_last_output())
    tex.print("\\end{lstlisting}")
end
\end{luacode}

\newcommand{\pytypeset}{\directlua{pytypeset()}}

\newenvironment{pyrepl}
{\PyLTVerbatimEnv\begin{pythonrepl}}
{\end{pythonrepl}\pytypeset}

\begin{document}

\begin{python}
def factorielle(n):
    fact = 1
    for k in range(n):
        fact = (k+1)*fact
    return fact	
\end{python}

\begin{pyrepl}
for _ in range(10):
	factorielle(5)

x = 1
\end{pyrepl}

\end{document}

```
the output of the function is printed at each iteration. It's not embarrassing but I think it's not the usual behavior of the REPL (see e.g. Spyder 5 below). 

<img width="277" alt="Capture d’écran 2021-11-21 à 18 19 13" src="https://user-images.githubusercontent.com/19306526/142772273-3d972226-166b-4c98-91a1-f9422824cb31.png">

**Reply by tndrle** [reply]
Thanks for the detailed example.

> but I think it's not the usual behavior of the REPL

Actually it depends on the environment. If you run the code in a pure Python REPL (start `python3` on the command line), then you see the same behavior as that of PyLuaTeX, i.e. repeated output.
